### DISCUSSION TITLE
Using Cforth, looking for serial examples

### DISCUSSION THREAD

**Discussion by jemo07** [ForthHub/discussion#152] [Q&A] [answered]
Hi all, 

I just discovered Mitch Bradley’s amazing Cforth, and I specially love the way he has created a VM with a pre compiled dictionary and a plethora of supported MCU’s. 

I can’t seem to find a lot of examples on how use this forth, lots of words, and would love if someone can point me in the right direction. I’m keen on simple UART serial communication on the RP2040 or the ESP32. 

Forth the EP2040, I think I have most of the code ready to enable the pins, but I don’t have all the details and most of the APP’s folder examples have a lot of missing little words that I can’t seem to find. https://gist.github.com/jemo07/d511aa14da7fc34e9e90bf4a0ecc390f

I’m mostly interested in learning forth, while building fun little projects just to test and showcase Cforth to the rest of the community.

Cheers!

**Comment by MitchBradley** [comment]
"Lots of missing little words" - tell us some specific examples.  I can't really address "lots of".

**Reply by jemo07** [reply]
Pleasure to greet you Mitch, apologies for the lack of clarity and concrete examples, let me work on providing exact details as I test the code, happy to collaborate in testing and documentation, I can commit to that. And again, thank you for gifting us with such a great tool!

**Comment by quozl** [comment]
https://github.com/MitchBradley/cforth/blob/master/README.md explains how to build for RP2040 using PlatformIO, which leverages the C++ UART implementation in the Arduino framework.  If you're wanting to reimplement a UART driver in Forth, you will have to disable that C++ implementation somehow, or steal the hardware out from under it.  Good luck!

**Reply by jemo07** [reply]
Hello Quozl, thank you, I have the code running on ESP32/STM32F4/and RP2040. Can you help me, where in the code can I find how that is been implemented, I'm happy to hunt and learn, that could be a great pointer in the right direction. To know that it's been implemented, saves me allot of work... I just to find how to enable a pin in serial mode for RX or TX...

**Reply by quozl** [reply]
When using PlatformIO to build, what you ask for isn't in C Forth.  Exactly where it is depends on the on the target board you are building for.  You could use a link map.

Or, assuming RP2040 , as a consequence of build you will have on your filesystem `.platformio/packages/framework-arduino-mbed` and within that a directory `cores/arduino` which contains
* [`Serial.cpp`](https://github.com/arduino/ArduinoCore-mbed/blob/main/cores/arduino/Serial.cpp) and
* [`Serial.h`](https://github.com/arduino/ArduinoCore-mbed/blob/main/cores/arduino/Serial.h)  for using the UART peripherals, or
* [`USB/USBSerial.cpp`](https://github.com/arduino/ArduinoCore-mbed/blob/main/cores/arduino/USB/USBSerial.cpp) and
* [`USB/USBCDC.cpp`](https://github.com/arduino/ArduinoCore-mbed/blob/main/cores/arduino/USB/USBCDC.cpp) for using the USB serial interface.

C Forth built using PlatformIO binds these in
* [`src/app/arduino/main.cpp`](https://github.com/MitchBradley/cforth/blob/master/src/app/arduino/main.cpp).

If there was a need to optimise to reduce external dependencies, a UART could be programmed directly.  Several C Forth build targets don't use PlatformIO at all, and have either direct UART programming or use other API such as FreeRTOS or platform SDKs.  C Forth hosted on a POSIX operating system uses the POSIX API for serial ports.

Links included above to the files in their GitHub repositories.

**Reply by jemo07** [reply]
Apologies for not clarifying better my request, understanding the build process, there must be a place in the code that one links the C API to forth words, this is what I was asking as it would allow me to work on a list, 

I did find such a place, but I was not able to find one that dressed this ( may be I might have overlooked it ) but I did manage to create a list of words, in this case in the ESP32 directory... 

https://gist.github.com/jemo07/6ed1af9df1f9ff28d5c9601d86ecc73b

**Reply by quozl** [reply]
Yes, and that's a list used by the ESP-IDK build of C Forth.  [`src/app/esp32/extend.c`](https://github.com/MitchBradley/cforth/blob/master/src/app/esp32/extend.c)

The PlatformIO build defines a different, shorter list.  [`src/platform/arduino/extend.c`](https://github.com/MitchBradley/cforth/blob/master/src/platform/arduino/extend.c)

Whichever build method you choose, you can further change these files to add C implementations of Forth words.

**Comment by MitchBradley** [comment]
There are example applications in various app/ subdirectories like arm-teensy3-pump, bottle-filler, esp8266-rtos-keg, filler, keg, labeler, and scale.  Some of them are for automating various processes in a Kombucha production kitchen.  The mqtt-* apps are IoT gadgets that are part of an IoT demo system that involved starting the demo with a sonoff switch, then using an RFID tag to start a robot car, using an alcohol sensor as a breathalyzer, using load cells from an old bathroom scale as an accelerator, sending speed commands to a robot car, and finally taking a selfie with an ESP32-CAM concealed inside a Bender (Futurama) action figure.  The individual components were ESP8266 and ESP32 based gadgets, and the storyboard ran on a tablet computer with NODE-Red managing overall various state transitions.

**Reply by MitchBradley** [reply]
The system was setup in the demo room at a Hacker's Conference, and I have a photo of Donald Knuth from the ESP32-CAM.

**Comment by MitchBradley** [comment, marked as answer]
cforth is not intended to be an all-inclusive product with an extensive fixed list of system functions.  Instead, it is whatever I or other need it to be for specific projects.  When I need additional functions in extend.c, I often create a new directory under src/app for the specific application, adjusting extend.c as needed.

**Comment by jemo07** [comment]
Hello Mitch, Yes I’m stating to comprehend that, thank you for bringing in some clarity.  Apologies for the lack of understanding. This has helped me a I can further evaluate how to get the serial communication needed. This brings a new level of clarity in understanding the app folders. This sets me in an exiting opportunity to better understand C-Forth. 

Thank you @MitchBradley and @quozl for shedding some light into this, will close this and comeback if and when I run to a standstill.
